### ISSUE TITLE
Trafic Prediction

### ISSUE THREAD

**Issue by manvi2330** [manvi2330/traffic-prediction#1]
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, r2_score
from datetime import datetime, timedelta

# Set random seed for reproducibility
np.random.seed(42)

# Generate time-series traffic data
num_samples = 96  # 24 hours at 15-minute intervals
time_series = [datetime(2025, 1, 16) + timedelta(minutes=15 * i) for i in range(num_samples)]

# Generate synthetic traffic metrics
data = pd.DataFrame({
    'time': time_series,
    'vehicle_count': np.random.randint(0, 100, num_samples),
    'traffic_speed': np.random.uniform(10, 60, num_samples),
    'congestion_level': np.random.uniform(0, 1, num_samples)
})

# Plot traffic data trends
plt.figure(figsize=(10, 6))
plt.plot(data['time'], data['vehicle_count'], label='Vehicle Count')
plt.plot(data['time'], data['traffic_speed'], label='Traffic Speed (km/h)')
plt.plot(data['time'], data['congestion_level'], label='Congestion Level')
plt.title('Simulated Traffic Data')
plt.xlabel('Time')
plt.ylabel('Values')
plt.legend()
plt.xticks(rotation=45)
plt.show()

# Feature selection for prediction
X = data[['vehicle_count', 'traffic_speed']]
y = data['congestion_level']

# Normalize features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3, random_state=42)

# Linear Regression model
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)
pred_lr = linear_model.predict(X_test)

# Evaluate Linear Regression model
print("Linear Regression Feature Weights:")
for feature, coef in zip(X.columns, linear_model.coef_):
    print(f"{feature}: {coef:.2f}")
print(f"\nLinear Regression - MSE: {mean_squared_error(y_test, pred_lr):.4f}, R²: {r2_score(y_test, pred_lr):.4f}")

# Random Forest Regressor model
rf_model = RandomForestRegressor(random_state=42)
rf_model.fit(X_train, y_train)
pred_rf = rf_model.predict(X_test)

# Evaluate Random Forest model
print(f"\nRandom Forest - MSE: {mean_squared_error(y_test, pred_rf):.4f}, R²: {r2_score(y_test, pred_rf):.4f}")

# Visualization of predictions
plt.figure(figsize=(10, 6))
plt.plot(y_test.values, label='Actual', color='b')
plt.plot(pred_lr, label='Linear Regression', linestyle='--', color='orange')
plt.plot(pred_rf, label='Random Forest', linestyle='-.', color='green')
plt.title('Congestion Prediction')
plt.xlabel('Time Interval')
plt.ylabel('Congestion Level')
plt.legend()
plt.show()

# Traffic signal control based on congestion predictions
green_durations = [30 if c < 0.3 else 60 if c < 0.7 else 90 for c in pred_rf]

# Plot signal durations
plt.figure(figsize=(10, 6))
plt.plot(green_durations, label='Green Light Duration (s)', color='g')
plt.title('Traffic Light Adjustments')
plt.xlabel('Time Interval')
plt.ylabel('Duration (seconds)')
plt.legend()
plt.show()

# Multi-intersection traffic control
num_intersections = 5
intersection_data = {
    f"Intersection_{i+1}": np.random.uniform(0.1, 0.9, len(pred_rf)) for i in range(num_intersections)
}

def determine_signal(congestion):
    return "Green: 30s" if congestion < 0.3 else "Green: 60s" if congestion < 0.7 else "Green: 90s"

print("\nTraffic Signal Adjustments for Intersections:")
for i in range(5):
    print(f"\nTime: {data['time'].iloc[i]}")
    for name, congestion_values in intersection_data.items():
        print(f"  {name}: Congestion = {congestion_values[i]:.2f}, Signal = {determine_signal(congestion_values[i])}")
